File: frontend/src/services/matchService.ts
import { get, post, patch, del } from './api';

// Types for matches (matching backend entity)
export interface Match {
  id: number;
  home_team_id?: number;
  away_team_id?: number;
  home_team_name?: string;
  away_team_name?: string;
  sport: string;
  location?: string;
  scheduled_at: string;
  status?: string;
  max_players?: number;
  description?: string;
  created_at?: string;
  updated_at?: string;
  // Relations (populated by backend) - can be string or object
  home_team?: string | {
    id: number;
    name: string;
  };
  away_team?: string | {
    id: number;
    name: string;
  };
  // Additional fields for UI
  teams?: string;
  time?: string;
  players?: string;
  price?: string;
}

export interface MatchParticipant {
  id?: number;
  match_id?: number;
  user_id: number;
  role?: string;
  goals?: number;
  assists?: number;
  rating?: number;
}

export interface MatchStat {
  id: number;
  match_id: number;
  user_id: number;
  goals?: number;
  assists?: number;
  rating?: number;
}

export interface CreateMatchData {
  home_team_id?: number;
  away_team_id?: number;
  sport: string;
  scheduled_at: string;
  location?: string;
  // Team names (for when not using team IDs)
  home_team?: string;
  away_team?: string;
  max_players?: number;
  description?: string;
}

export interface UpdateMatchStatusData {
  status: string;
}

// Match API functions
export const matchService = {
  // Get all matches
  async getMatches(params?: Record<string, string>): Promise<Match[]> {
    return get<Match[]>('/api/matches', params);
  },

  // Get a single match by ID
  async getMatch(id: number): Promise<Match> {
    return get<Match>(`/api/matches/${id}`);
  },

  // Create a new match
  async createMatch(data: CreateMatchData): Promise<Match> {
    return post<Match>('/api/matches', data);
  },

  // Add a participant to a match
  async addParticipant(matchId: number, data: MatchParticipant): Promise<void> {
    return post(`/api/matches/${matchId}/participants`, data);
  },

  // Update match stats for a participant
  async updateMatchStats(matchId: number, data: MatchParticipant): Promise<void> {
    return post(`/api/matches/${matchId}/stats`, data);
  },

  // Update match status
  async updateMatchStatus(matchId: number, data: UpdateMatchStatusData): Promise<Match> {
    return patch<Match>(`/api/matches/${matchId}/status`, data);
  },

  // Delete a match
  async deleteMatch(id: number): Promise<void> {
    return del(`/api/matches/${id}`);
  },
};

export default matchService;

File: rival_mobile/lib/services/match_service.dart
// Match Service
// Mirrors frontend/src/services/matchService.ts functionality

import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:rival_mobile/models/match_models.dart';
import 'package:rival_mobile/services/api_client.dart';

/// Service for Match-related API calls
class MatchService {
  final ApiClient _apiClient;

  MatchService(this._apiClient);

  Future<List<Match>> getMatches() async {
    return await _apiClient.get<List<Match>>('/matches');
  }

  Future<Match> getMatch(int id) async {
    return await _apiClient.get<Match>('/matches/$id');
  }

  Future<Match> createMatch({
    int? homeTeamId,
    int? awayTeamId,
    required String sport,
    required String scheduledAt,
    String? location,
  }) async {
    return await _apiClient.post<Match>(
      '/matches',
      data: CreateMatchData(
        homeTeamId: homeTeamId,
        awayTeamId: awayTeamId,
        sport: sport,
        scheduledAt: scheduledAt,
        location: location,
      ).toJson(),
    );
  }

  Future<void> addParticipant({
    required int matchId,
    required int userId,
    String? role,
  }) async {
    await _apiClient.post(
      '/matches/$matchId/participants',
      data: {
        'user_id': userId,
        'role': role,
      },
    );
  }

  Future<void> updateMatchStats({
    required int matchId,
    required int userId,
    int? goals,
    int? assists,
    double? rating,
  }) async {
    await _apiClient.post(
      '/matches/$matchId/stats',
      data: {
        'user_id': userId,
        'goals': goals,
        'assists': assists,
        'rating': rating,
      },
    );
  }

  Future<Match> updateMatchStatus({
    required int matchId,
    required String status,
  }) async {
    return await _apiClient.patch<Match>(
      '/matches/$matchId/status',
      data: UpdateMatchStatusData(status: status).toJson(),
    );
  }

  Future<void> deleteMatch(int id) async {
    await _apiClient.delete('/matches/$id');
  }
}

// Provider for MatchService
final matchServiceProvider = Provider<MatchService>((ref) {
  final apiClient = ref.watch(apiClientProvider);
  return MatchService(apiClient);
});

// Match State Notifier using Riverpod
class MatchState {
  final List<Match> matches;
  final bool isLoading;
  final String? error;

  const MatchState({
    this.matches = const [],
    this.isLoading = false,
    this.error,
  });

  const MatchState.initial()
      : matches = const [],
        isLoading = false,
        error = null;

  MatchState copyWith({
    List<Match>? matches,
    bool? isLoading,
    String? error,
  }) {
    return MatchState(
      matches: matches ?? this.matches,
      isLoading: isLoading ?? this.isLoading,
      error: error ?? this.error,
    );
  }
}

class MatchNotifier extends StateNotifier<MatchState> {
  final MatchService _matchService;

  MatchNotifier(this._matchService) : super(const MatchState.initial());

  Future<void> fetchMatches() async {
    state = state.copyWith(isLoading: true, error: null);
    try {
      final matches = await _matchService.getMatches();
      state = state.copyWith(
        matches: matches,
        isLoading: false,
      );
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        error: e.toString(),
      );
    }
  }

  Future<Match> createMatch(CreateMatchData data) async {
    state = state.copyWith(isLoading: true, error: null);
    try {
      final match = await _matchService.createMatch(
        homeTeamId: data.homeTeamId,
        awayTeamId: data.awayTeamId,
        sport: data.sport,
        scheduledAt: data.scheduledAt,
        location: data.location,
      );
      state = state.copyWith(
        matches: [...state.matches, match],
        isLoading: false,
      );
      return match;
    } catch (e) {
      state = state.copyWith(
        isLoading: false,
        error: e.toString(),
      );
      rethrow;
    }
  }

  Future<void> joinMatch({
    required int matchId,
    required int userId,
  }) async {
    try {
      await _matchService.addParticipant(
        matchId: matchId,
        userId: userId,
      );
      // Refresh matches after joining
      await fetchMatches();
    } catch (e) {
      state = state.copyWith(error: e.toString());
      rethrow;
    }
  }
}

// Provider for MatchNotifier
final matchNotifierProvider = StateNotifierProvider<MatchNotifier, MatchState>((ref) {
  final matchService = ref.watch(matchServiceProvider);
  return MatchNotifier(matchService);
});
